### ISSUE TITLE
Possible gradient miscalculation with repeated indices in Autograd Functions `PutBackward0`, `IndexPutBackward0` and `ScatterBackward0`

### ISSUE THREAD

**Issue by mntsx** [pytorch/pytorch#146299]
### 🐛 Describe the bug

I think this is an incorrect behaviour for `PutBackward0`, `IndexPutBackward0` and `ScatterBackward0`:

```python
>>> import torch
>>> T = torch.zeros(size=(1,), requires_grad=True)
>>> idx = torch.tensor([0,0], dtype=torch.long)
>>> src = torch.tensor([1.0, 2.0], requires_grad=True)
>>> O = torch.put(input=T, index=idx, source=src)
>>> O = O.sum()
>>> O.backward()
>>> print(O)
tensor(2., grad_fn=<SumBackward0>)
>>> print(src.grad)
tensor([1., 1.])
```

```python
>>> import torch
>>> T = torch.zeros(size=(1,), requires_grad=True)
>>> idx = torch.tensor([0,0], dtype=torch.long)
>>> src = torch.tensor([1.0, 2.0], requires_grad=True)
>>> C = T.clone()
>>> C[idx] = src
>>> O = C.sum()
>>> O.backward()
>>> print(O)
tensor(2., grad_fn=<SumBackward0>)
>>> print(src.grad)
tensor([1., 1.])
```

```python
>>> import torch
>>> T = torch.zeros(size=(1,), requires_grad=True)
>>> idx = torch.tensor([0, 0], dtype=torch.long)
>>> src = torch.tensor([1.0, 2.0], requires_grad=True)
>>> O = torch.scatter(input=T, dim=0, index=idx, src=src)
>>> O = O.sum()
>>> O.backward()
>>> print(O)
tensor(2., grad_fn=<SumBackward0>)
>>> print(src.grad)
tensor([1., 1.])
```

If elements inserted at the same index overwrite each other (and the output being `2.` instead of `3.` shows that they do overwrite each other), then `src.grad[0]` should be `0.`, not `1.`.

### Versions

```bash
python collect_env.py
--2025-02-03 00:54:07--  https://raw.githubusercontent.com/pytorch/pytorch/main/torch/utils/collect_env.py
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.109.133, 185.199.108.133, 185.199.111.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.109.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 24353 (24K) [text/plain]
Saving to: ‘collect_env.py.1’

collect_env.py.1              100%[=================================================>]  23.78K  --.-KB/s    in 0.005s

2025-02-03 00:54:07 (5.15 MB/s) - ‘collect_env.py.1’ saved [24353/24353]

/home/miguelmnts/.pyenv/versions/3.12.6/lib/python3.12/site-packages/torch/_subclasses/functional_tensor.py:295: UserWarning: Failed to initialize NumPy: No module named 'numpy' (Triggered internally at ../torch/csrc/utils/tensor_numpy.cpp:84.)
  cpu = _conversion_method_template(device=torch.device("cpu"))

Collecting environment information...
PyTorch version: 2.5.1+cu124
Is debug build: False
CUDA used to build PyTorch: 12.4
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.3 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.35

Python version: 3.12.6 (main, Sep 25 2024, 17:04:58) [GCC 11.4.0] (64-bit runtime)
Python platform: Linux-5.15.133.1-microsoft-standard-WSL2-x86_64-with-glibc2.35
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                       x86_64
CPU op-mode(s):                     32-bit, 64-bit
Address sizes:                      48 bits physical, 48 bits virtual
Byte Order:                         Little Endian
CPU(s):                             8
On-line CPU(s) list:                0-7
Vendor ID:                          AuthenticAMD
Model name:                         AMD Ryzen 5 7520U with Radeon Graphics
CPU family:                         23
Model:                              160
Thread(s) per core:                 2
Core(s) per socket:                 4
Socket(s):                          1
Stepping:                           0
BogoMIPS:                           5588.57
Flags:                              fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl tsc_reliable nonstop_tsc cpuid extd_apicid pni pclmulqdq ssse3 fma cx16 sse4_1 sse4_2 movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm cmp_legacy svm cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw topoext perfctr_core ssbd ibrs ibpb stibp vmmcall fsgsbase bmi1 avx2 smep bmi2 rdseed adx smap clflushopt clwb sha_ni xsaveopt xsavec xgetbv1 clzero xsaveerptr arat npt nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold v_vmsave_vmload umip rdpid
Virtualization:                     AMD-V
Hypervisor vendor:                  Microsoft
Virtualization type:                full
L1d cache:                          128 KiB (4 instances)
L1i cache:                          128 KiB (4 instances)
L2 cache:                           2 MiB (4 instances)
L3 cache:                           4 MiB (1 instance)
Vulnerability Gather data sampling: Not affected
Vulnerability Itlb multihit:        Not affected
Vulnerability L1tf:                 Not affected
Vulnerability Mds:                  Not affected
Vulnerability Meltdown:             Not affected
Vulnerability Mmio stale data:      Not affected
Vulnerability Retbleed:             Mitigation; untrained return thunk; SMT enabled with STIBP protection
Vulnerability Spec rstack overflow: Mitigation; safe RET
Vulnerability Spec store bypass:    Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:           Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:           Mitigation; Retpolines, IBPB conditional, STIBP always-on, RSB filling, PBRSB-eIBRS Not affected
Vulnerability Srbds:                Not affected
Vulnerability Tsx async abort:      Not affected

Versions of relevant libraries:
[pip3] nvidia-cublas-cu12==12.4.5.8
[pip3] nvidia-cuda-cupti-cu12==12.4.127
[pip3] nvidia-cuda-nvrtc-cu12==12.4.127
[pip3] nvidia-cuda-runtime-cu12==12.4.127
[pip3] nvidia-cudnn-cu12==9.1.0.70
[pip3] nvidia-cufft-cu12==11.2.1.3
[pip3] nvidia-curand-cu12==10.3.5.147
[pip3] nvidia-cusolver-cu12==11.6.1.9
[pip3] nvidia-cusparse-cu12==12.3.1.170
[pip3] nvidia-nccl-cu12==2.21.5
[pip3] nvidia-nvjitlink-cu12==12.4.127
[pip3] nvidia-nvtx-cu12==12.4.127
[pip3] torch==2.5.1
[pip3] torch_nd_conv==0.1.0
[pip3] triton==3.1.0
[conda] Could not collect
```

cc @ezyang @albanD @gqchen @pearu @nikitaved @soulitzer @Varal7 @xmfan

**Comment by malfet** [comment]
This feels like an expected behavior per [`index_put_`](https://pytorch.org/docs/stable/generated/torch.Tensor.index_put_.html#torch.Tensor.index_put_) documentation:
 > If accumulate is `False`, the behavior is undefined if indices contain duplicate elements.

**Comment by mntsx** [comment]
Okay, I thought that statement referred to the fact that when there are overlapping indices, the elements get overwritten, not that the gradient calculation was incorrect according to that methodology.

**Comment by albanD** [comment]
Since we don't know which one was chosen during the forward, we can't really provide corresponding gradients easily. Any technique to try and guess + do backward mapping for this would be significantly slower.

**Comment by mntsx** [comment]
Maybe I'm wrong, but as far as I can see, it is indeed known that the last indexed element with the repeated index is the one that gets selected. I understand that checking the uniqueness of each index can be costly. Additionally, the current behavior is not necessarily undesirable if it is assumed that repeated indices always correspond to identical elements in the source, since in that case, overwriting is indistinguishable from superposition. Nevertheless, it would probably be worth being more explicit about this behavior in the corresponding module's documentation.

**Comment by albanD** [comment]
> Maybe I'm wrong, but as far as I can see, it is indeed known that the last indexed element with the repeated index is the one that gets selected. 

That depends on the device and input size. In particular for larger inputs, once the operation get split among multiple threads, any of them could be selected.

> Nevertheless, it would probably be worth being more explicit about this behavior in the corresponding module's documentation.

I 100% agree that while "undefined behavior" is easy for us, it's not very helpful for users haha
We can definitely update that to mention that one will get selected arbitrarily and gradients will flow as if accumulate was set to True.
